engineioxide/src/payload.rs

```diff
@@ -0,0 +1,164 @@
+use std::{io::BufRead, vec};
+use cfg_if::cfg_if;
+
+use crate::protocol::ProtocolVersion;
+
+pub const PACKET_SEPARATOR: u8 = b'\x1e';
+
+/// A payload is a series of encoded packets tied together.
+/// How packets are tied together depends on the protocol.
+pub struct Payload<R: BufRead> {
+    reader: R,
+    buffer: Vec<u8>,
+    #[allow(dead_code)]
+    protocol: ProtocolVersion,
+}
+
+type Item = Result<String, String>;
+
+impl<R: BufRead> Payload<R> {
+    pub fn new(protocol: ProtocolVersion, data: R) -> Self {
+        Payload {
+            reader: data,
+            buffer: vec![],
+            protocol,
+        }
+    }
+
+    #[cfg(feature = "v3")]
+    fn next_v3(&mut self) -> Option<Item> {
+        self.buffer.clear();
+
+        match self.reader.read_until(b':', &mut self.buffer) {
+            Ok(bytes_read) => {
+                if bytes_read > 0 {
+                    // remove trailing separator
+                    if self.buffer.ends_with(&[b':']) {
+                        self.buffer.pop();
+                    }
+
+                    let length = match String::from_utf8(self.buffer.clone()) {
+                        Ok(s) => {
+                            if let Ok(l) = s.parse::<usize>() {
+                                l
+                            } else {
+                                return Some(Err("Invalid packet length".into()));
+                            }
+                        },
+                        Err(_) => return Some(Err("Invalid packet length".into())),
+                    };
+
+                    self.buffer.clear();
+                    self.buffer.resize(length, 0);
+
+                    match self.reader.read_exact(&mut self.buffer) {
+                        Ok(_) => {
+                            match String::from_utf8(self.buffer.clone()) {
```

Comment:
Cloning the buffer may have non significant impacts on packet parsing. I think it would be better to like this:
```rust
let buffer = std::mem::take(&mut self.buffer);
match String::from_utf8(buffer) {
```
Thanks to that it wouldn't be necessary to call `self.buffer.clear()`